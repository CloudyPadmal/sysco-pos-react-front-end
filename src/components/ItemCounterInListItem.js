import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import {
    dispatchQuickDeleteThisItemFromQuantity,
    dispatchQuickIndecrementItemFromQuantity
} from "../actions/ordercontrolactions";

/***************************************************************************************************
 *  Custom component to increment and decrement value within a minimum and
 * maximum range. A local count will be maintained to help with validation
 * @param name Identifier for this counter
 * @param min Minimum allowable value
 * @param max Maximum allowable value
 **************************************************************************************************/
class ItemCounterInListItem extends Component {

    /***********************************************************************************************
     * 
     **********************************************************************************************/
    deleteThisItem = (e) => {
        e.preventDefault();
        let P = this.props.itemsList.find((I) => I.productID === this.props.name).price;
        this.props.dispatchQuickDeleteThisItemFromQuantity(
            this.props.name, this.props.currentOrderID, this.props.count, P, this.props.passKey);
    }

    /***********************************************************************************************
     * 
     **********************************************************************************************/
    incrementItemCount = (e) => {
        e.preventDefault();
        if (this.props.count < this.props.max) {
            let P = this.props.itemsList.find((I) => I.productID === this.props.name).price;
            this.props.dispatchQuickIndecrementItemFromQuantity(
                this.props.name, true, this.props.count, this.props.currentOrderID,
                P, this.props.passKey);
        }
    }

    /***********************************************************************************************
     * 
     **********************************************************************************************/
    decrementItemCount = (e) => {
        e.preventDefault();
        if (this.props.count > this.props.min) {
            let P = this.props.itemsList.find(
                (I) => I.productID === this.props.name).price;
            this.props.dispatchQuickIndecrementItemFromQuantity(this.props.name, false,
                this.props.count, this.props.currentOrderID, P, this.props.passKey);
        }
    }

    render() {
        return (
            <React.Fragment>
                <div className="btn-toolbar" role="toolbar" aria-label="Toolbar with button groups">
                    <div className="btn-group" role="group" aria-label="Item-Group">
                        <button type="button" className="btn btn-secondary"
                            onClick={this.incrementItemCount}>&#43;</button>
                        <div style={{ width: "3em", textAlign: "center" }}>
                            <span className="input-group-text" style={{
                                display: "inline-block", borderRadius: "0px",
                                width: "100%", height: "100%"
                            }}>{this.props.count}</span></div>
                        <button type="button" className="btn btn-secondary"
                            onClick={this.decrementItemCount}>&#8722;</button>
                    </div>
                    <div className="btn-group"
                        role="group" aria-label="Delete-Group" style={{ marginLeft: "10px" }}>
                        <button
                            type="button" className="btn btn-danger"
                            onClick={this.deleteThisItem}><i className="fas fa-trash"></i></button>
                    </div>
                </div>
            </React.Fragment>
        );
    }
}

ItemCounterInListItem.propTypes = {
    dispatchQuickIndecrementItemFromQuantity: PropTypes.func.isRequired,
    dispatchQuickDeleteThisItemFromQuantity: PropTypes.func.isRequired
};

const mapStateToProps = (state) => ({
    passKey: state.uac.passKey,
    itemsList: state.ord.itemsList,
    currentOrderID: state.ord.currentOrderID
});

export default connect(mapStateToProps, {
    dispatchQuickIndecrementItemFromQuantity,
    dispatchQuickDeleteThisItemFromQuantity
})(ItemCounterInListItem);
