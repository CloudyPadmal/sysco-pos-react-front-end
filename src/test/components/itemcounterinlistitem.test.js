import React from "react";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { Provider } from "react-redux";
import { mount } from "../enzyme";
import { BrowserRouter as Router } from "react-router-dom";
import { CookiesProvider } from "react-cookie";

import ItemCounterInListItem from "../../components/ItemCounterInListItem";

const middlewares = [thunk];

describe("<ItemCounterInListItem /> component", () => {

    const increment = jest.fn();
    const deleteitem = jest.fn();

    const initialState = {
        uac: {
            isLoggedIn: true,
            passKey: "abcd"
        },
        ord: {
            currentOrderID: "abcdefg111",
            itemQuantity: { "A": 1, "B": 2, "C": 3 },
            itemsList: [
                { _id: "a", productID: "A", price: 1 },
                { _id: "b", productID: "B", price: 2 },
                { _id: "c", productID: "C", price: 3 }
            ],
            total: 100,
            currentOrder: {
                _id: "asdadasdasda",
                cartID: "abc",
                items: [
                    {}
                ]
            }
        },
        dispatch_QuickIndecrementItemFromQuantity: increment,
        dispatch_QuickDeleteThisItemFromQuantity: deleteitem
    };
    const mockStore = configureStore(middlewares);
    let store;

    beforeEach(() => {
        store = mockStore(initialState);
    });

    it("renders with correct value", () => {
        const wrapper = mount(
            <CookiesProvider>
                <Provider store={store}>
                    <Router>
                        <ItemCounterInListItem
                            name={"A"}
                            min={1}
                            max={10}
                            quantity={3} />
                    </Router>
                </Provider>
            </CookiesProvider>
        );
        expect(wrapper).toMatchSnapshot();
    });
    it("calls delete", () => {
        const wrapper = mount(
            <CookiesProvider>
                <Provider store={store}>
                    <Router>
                        <ItemCounterInListItem
                            name={"A"}
                            min={1}
                            max={10}
                            quantity={3} />
                    </Router>
                </Provider>
            </CookiesProvider>
        );
        wrapper.find(".btn-danger").simulate("click");
        expect(wrapper.contains(<i className="fas fa-trash" />)).toBeTruthy();
    });
    it("calls increment", () => {
        const wrapper = mount(
            <CookiesProvider>
                <Provider store={store}>
                    <Router>
                        <ItemCounterInListItem
                            name={"A"}
                            count={5}
                            min={1}
                            max={10}
                            quantity={3} />
                    </Router>
                </Provider>
            </CookiesProvider>
        );
        wrapper.find(".btn-secondary").at(0).simulate("click");
        expect(wrapper.contains("+")).toBeTruthy();
    });
    it("calls increment blocked", () => {
        const wrapper = mount(
            <CookiesProvider>
                <Provider store={store}>
                    <Router>
                        <ItemCounterInListItem
                            name={"A"}
                            count={10}
                            min={1}
                            max={10}
                            quantity={3} />
                    </Router>
                </Provider>
            </CookiesProvider>
        );
        wrapper.find(".btn-secondary").at(0).simulate("click");
        expect(wrapper.contains("+")).toBeTruthy();
    });
    it("calls decrement", () => {
        const wrapper = mount(
            <CookiesProvider>
                <Provider store={store}>
                    <Router>
                        <ItemCounterInListItem
                            name={"A"}
                            count={5}
                            min={1}
                            max={10}
                            quantity={3} />
                    </Router>
                </Provider>
            </CookiesProvider>
        );
        wrapper.find(".btn-secondary").at(1).simulate("click");
        expect(wrapper.contains("−")).toBeTruthy();
    });

    it("calls decrement blocked", () => {
        const wrapper = mount(
            <CookiesProvider>
                <Provider store={store}>
                    <Router>
                        <ItemCounterInListItem
                            name={"A"}
                            count={1}
                            min={1}
                            max={10}
                            quantity={3} />
                    </Router>
                </Provider>
            </CookiesProvider>
        );
        wrapper.find(".btn-secondary").at(1).simulate("click");
        expect(wrapper.contains("−")).toBeTruthy();
    });
});