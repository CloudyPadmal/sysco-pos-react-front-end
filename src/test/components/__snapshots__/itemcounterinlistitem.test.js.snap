// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItemCounterInListItem /> component renders with correct value 1`] = `
<CookiesProvider>
  <Provider
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <BrowserRouter>
      <Router
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "createHref": [Function],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        }
      >
        <Connect(ItemCounterInListItem)
          max={10}
          min={1}
          name="A"
          quantity={3}
        >
          <ItemCounterInListItem
            currentOrderID="abcdefg111"
            dispatchQuickDeleteThisItemFromQuantity={[Function]}
            dispatchQuickIndecrementItemFromQuantity={[Function]}
            itemsList={
              Array [
                Object {
                  "_id": "a",
                  "price": 1,
                  "productID": "A",
                },
                Object {
                  "_id": "b",
                  "price": 2,
                  "productID": "B",
                },
                Object {
                  "_id": "c",
                  "price": 3,
                  "productID": "C",
                },
              ]
            }
            max={10}
            min={1}
            name="A"
            passKey="abcd"
            quantity={3}
          >
            <div
              aria-label="Toolbar with button groups"
              className="btn-toolbar"
              role="toolbar"
            >
              <div
                aria-label="Item-Group"
                className="btn-group"
                role="group"
              >
                <button
                  className="btn btn-secondary"
                  onClick={[Function]}
                  type="button"
                >
                  +
                </button>
                <div
                  style={
                    Object {
                      "textAlign": "center",
                      "width": "3em",
                    }
                  }
                >
                  <span
                    className="input-group-text"
                    style={
                      Object {
                        "borderRadius": "0px",
                        "display": "inline-block",
                        "height": "100%",
                        "width": "100%",
                      }
                    }
                  />
                </div>
                <button
                  className="btn btn-secondary"
                  onClick={[Function]}
                  type="button"
                >
                  âˆ’
                </button>
              </div>
              <div
                aria-label="Delete-Group"
                className="btn-group"
                role="group"
                style={
                  Object {
                    "marginLeft": "10px",
                  }
                }
              >
                <button
                  className="btn btn-danger"
                  onClick={[Function]}
                  type="button"
                >
                  <i
                    className="fas fa-trash"
                  />
                </button>
              </div>
            </div>
          </ItemCounterInListItem>
        </Connect(ItemCounterInListItem)>
      </Router>
    </BrowserRouter>
  </Provider>
</CookiesProvider>
`;
